runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/commits/StreamingEngineWorkCommitter.java

```diff
@@ -0,0 +1,210 @@
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.beam.runners.dataflow.worker.streaming.WeightedBoundedQueue;
+import org.apache.beam.runners.dataflow.worker.streaming.Work;
+import org.apache.beam.runners.dataflow.worker.windmill.client.CloseableStream;
+import org.apache.beam.runners.dataflow.worker.windmill.client.WindmillStream.CommitWorkStream;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Streaming engine implementation of {@link WorkCommitter}. Commits work back to Streaming Engine
+ * backend.
+ */
+@Internal
+@ThreadSafe
+public final class StreamingEngineWorkCommitter implements WorkCommitter {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingEngineWorkCommitter.class);
+  private static final int COMMIT_BATCH_SIZE = 5;
+  private static final int TARGET_COMMIT_BATCH_SIZE = 500 << 20; // 500MB
+
+  private final Supplier<CloseableStream<CommitWorkStream>> commitWorkStreamFactory;
+  private final WeightedBoundedQueue<Commit> commitQueue;
+  private final ExecutorService commitSenders;
+  private final AtomicLong activeCommitBytes;
+  private final Consumer<CompleteCommit> onCommitComplete;
+  private final int numCommitSenders;
+
+  private StreamingEngineWorkCommitter(
+      Supplier<CloseableStream<CommitWorkStream>> commitWorkStreamFactory,
+      int numCommitSenders,
+      Consumer<CompleteCommit> onCommitComplete) {
+    this.commitWorkStreamFactory = commitWorkStreamFactory;
+    this.commitQueue =
+        WeightedBoundedQueue.create(
+            TARGET_COMMIT_BATCH_SIZE,
+            commit -> Math.min(TARGET_COMMIT_BATCH_SIZE, commit.getSize()));
+    this.commitSenders =
+        Executors.newFixedThreadPool(
+            numCommitSenders,
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setPriority(Thread.MAX_PRIORITY)
+                .setNameFormat("CommitThread-%d")
+                .build());
+    this.activeCommitBytes = new AtomicLong();
+    this.onCommitComplete = onCommitComplete;
+    this.numCommitSenders = numCommitSenders;
+  }
+
+  public static StreamingEngineWorkCommitter create(
+      Supplier<CloseableStream<CommitWorkStream>> commitWorkStreamFactory,
+      int numCommitSenders,
+      Consumer<CompleteCommit> onCommitComplete) {
+    return new StreamingEngineWorkCommitter(
+        commitWorkStreamFactory, numCommitSenders, onCommitComplete);
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public void start() {
+    if (!commitSenders.isShutdown() || !commitSenders.isTerminated()) {
```

Comment:
seems like should be &&, or just have the shutdown check since I think it has to precede termination.

Reply:
done